homeassistant/components/unifiprotect/models.py

```diff
@@ -21,7 +21,7 @@ class ProtectRequiredKeysMixin:
 
     ufp_required_field: str | None = None
     ufp_value: str | None = None
-    ufp_value_fn: Callable[[ProtectAdoptableDeviceModel | NVR], Any] | None = None
+    ufp_value_fn: Callable[[Any], Any] | None = None
```

Comment:
I don't agree `Any` is a solution. If anything, it should become more specific, and less generic.

Reply:
mypy does not like the more specific one. Even though all of the methods use either these two or child classes of them. let me see if using a base class shared by `ProtectAdoptableDeviceModel` and `NVR` works. Maybe that is the issue.

Ref: https://github.com/home-assistant/core/pull/63716#discussion_r780820579

Reply:
> mypy does not like the more specific one

That means typing is incorrect; that needs to be solved, not suppressed.

Reply:
Yeah, using the better base class works. It is annoying though because it was not incorrect. `Camera` is a subclass of `ProtectAdoptableDeviceModel` so it fit in the original one. I guess you just cannot do unions and then remove them further down the line?

Reply:
Nope, nevermind, that brings back all of the mypy errors. They were just not showing up in the pre-commit hook.

Reply:
I played with it a bit and it looks like it needs a mix-in per device type to avoid the Any.

Reply:
Yeah, which would mean up to 1-6 EntityDescriptions per platform and would still need the base mixin to be Any to make mypy happy without duplicating the dataclasses for every device type.

Reply:
> I played with it a bit and it looks like it needs a mix-in per device type to avoid the Any.

That is correct.

> I really do think Any is the best solution here.

Any is never a solution. It is suppressing.
If the upstream library is typed, we should leverage that.

Reply:
The upstream library is typed. mypy is just not allowing to make a function with a more specific argument when you know it is going to be unless the original is `Any`. 

I would love to get rid of Any but I cannot figure out how without making mypy complain in some way. I made a simple as possible reproduction case. Copy/paste this to the root of the HA root and then `mypy test.py` it.

https://gist.github.com/AngellusMortis/3242a3c75fbea54c6f827e4048bfe34b

If you have a solution that does not require one of the following, I will implement it right away:

* The use of an `Any`
* Asserting the type in the callables
* Using a type: ignore
* Making a different `TestClass` for _every_ combination

The `.device` attribute on the Entity class and the `_fn` attributes on the EntityDescription class are both fixed at class creation, and they are guaranteed to have their types match up. So, there is zero risk of a runtime error. I just do not know how to model that with mypy without making a Entity class and EntityDescription class for _every_ combination, which would add an additional 14 class (7 Entity and 7 EntityDescription) to make the existing code work.

Reply:
Can you try to use a `Generic`? I tried it on `switch.py` and it worked...

```python
T = TypeVar("T")

@dataclass
class ProtectRequiredKeysMixin(Generic[T]):
   ...

@dataclass
class ProtectSetableKeysMixin(ProtectRequiredKeysMixin, Generic[T]):
    """Mixin for settable values."""
```

```python
from .models import T, ProtectSetableKeysMixin

@dataclass
class ProtectSwitchEntityDescription(
    ProtectSetableKeysMixin, SwitchEntityDescription, Generic[T]
):
    """Describes UniFi Protect Switch entity."""
   ...

    ProtectSwitchEntityDescription[Camera](
        key="high_fps",
        name="High FPS",
        icon="mdi:video-high-definition",
        entity_category=EntityCategory.CONFIG,
        ufp_required_field="feature_flags.has_highfps",
        ufp_value_fn=_get_is_highfps,
        ufp_set_method_fn=_set_highfps,
    ),
```

Reply:
Thanks @epenet, I just looked at your PR. It looks great. I always forget you can create generics most of the time since I still use them so infrequently. I will pull in your PR and then update this one to clean everything up.